pkg/vpcmodel/grouping.go

```diff
@@ -0,0 +1,215 @@
+package vpcmodel
+
+import (
+	"sort"
+	"strings"
+)
+
+type ConnStr string
+
+type GroupingConnections struct {
+	publicConnectivity map[Node]map[ConnStr][]Node // for each line here can group list of external nodes to cidrs list as of one element
+}
+
+func (g *GroupingConnections) getGroupedConnLines(isSrcToDst bool) []*GroupedConnLine {
+	res := []*GroupedConnLine{}
+	for a, aMap := range g.publicConnectivity {
+		for conn, b := range aMap {
+			var resElem *GroupedConnLine
+			if isSrcToDst {
+				resElem = &GroupedConnLine{a, &GroupedExternalNodes{b}, conn}
+			} else {
+				resElem = &GroupedConnLine{&GroupedExternalNodes{b}, a, conn}
+			}
+			res = append(res, resElem)
+		}
+	}
+	return res
+}
+
+func newGroupingConnections() *GroupingConnections {
+	return &GroupingConnections{
+		publicConnectivity: map[Node]map[ConnStr][]Node{},
+	}
+}
+
+func newGroupConnLines(c *CloudConfig, v *VPCConnectivity) *GroupConnLines {
+	res := &GroupConnLines{c: c, v: v, srcToDst: newGroupingConnections(), DstToSrc: newGroupingConnections()}
+	res.computeGrouping()
+	return res
+}
+
+type GroupConnLines struct {
+	c            *CloudConfig
+	v            *VPCConnectivity
+	srcToDst     *GroupingConnections
+	DstToSrc     *GroupingConnections
+	GroupedLines []*GroupedConnLine
+}
+
+// EndpointElem can be Node(networkInterface) / NodeSet(subnet) / groupedExternalNodes
+type EndpointElem interface {
+	Name() string
+}
+
+type GroupedConnLine struct {
+	Src  EndpointElem
+	Dst  EndpointElem
+	Conn ConnStr
+}
+
+func (g *GroupedConnLine) String() string {
+	return g.Src.Name() + " => " + g.Dst.Name() + " : " + string(g.Conn)
+}
+
+func (g *GroupedConnLine) getSrcOrDst(isSrc bool) EndpointElem {
+	if isSrc {
+		return g.Src
+	}
+	return g.Dst
+}
+
+// implements endpointElem interface
+type GroupedExternalNodes struct {
+	nodes []Node
+}
+
+func (g *GroupedExternalNodes) Name() string {
+	isAllIntenetRange, err := isEntirePublicInternetRange(g.nodes)
+	if err == nil && isAllIntenetRange {
+		return "Public Internet (all ranges)"
+	}
+	nodesStrings := make([]string, len(g.nodes))
+	for i, n := range g.nodes {
+		nodesStrings[i] = n.Cidr()
+	}
+	sort.Strings(nodesStrings)
+	return strings.Join(nodesStrings, ",")
+}
+
+func (g *GroupingConnections) addPublicConnectivity(n Node, conn ConnStr, target Node) {
+	if _, ok := g.publicConnectivity[n]; !ok {
+		g.publicConnectivity[n] = map[ConnStr][]Node{}
+	}
+	if _, ok := g.publicConnectivity[n][conn]; !ok {
+		g.publicConnectivity[n][conn] = []Node{}
+	}
+	g.publicConnectivity[n][conn] = append(g.publicConnectivity[n][conn], target)
+}
+
+func subnetGrouping(elemsList []EndpointElem, c *CloudConfig) []EndpointElem {
```

Comment:
This function needs more documentation. It's not trivial

Reply:
done